src/Data/List/Linear.hs

```diff
@@ -0,0 +1,346 @@
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+
+{-|
+Linear versions of 'Data.List' functions.
+
+This module only contains minimal amount of documentation; consult the
+original "Data.List" module for more detailed information.
+-}
+module Data.List.Linear
+  ( -- * Basic functions
+    (++)
+  , map
+  , filter
+  , NonLinear.head
+  , uncons
+  , NonLinear.tail
+  , NonLinear.last
+  , NonLinear.init
+  , reverse
+  , length
+  , NonLinear.null
+    -- * Extracting sublists
+  , take
+  , drop
+  , splitAt
+  , span
+  , partition
+  , takeWhile
+  , dropWhile
+  , NonLinear.find
+  , intersperse
+  , intercalate
+  , transpose
+  -- * Folds
+  , foldl
+  , foldl'
+  , foldl1
+  , foldl1'
+  , foldr
+  , foldr1
+  , foldMap
+  , foldMap'
+  -- * Special folds
+  , concat
+  , concatMap
+  , and
+  , or
+  , any
+  , all
+  , sum
+  , product
+  -- * Building lists
+  , scanl
+  , scanl1
+  , scanr
+  , scanr1
+  , repeat
+  , replicate
+  , cycle
+  , iterate
+  , unfoldr
+  -- * Zipping lists
+  , zip
+  , zip'
+  , zip3
+  , zipWith
+  , zipWith'
+  , zipWith3
+  , unzip
+  , unzip3
+  ) where
+
+import qualified Unsafe.Linear as Unsafe
+import qualified Prelude as Prelude
+import Prelude (Maybe(..), Either(..), Int, otherwise)
+import Prelude.Linear.Internal
+import Data.Bool.Linear
+import Data.Unrestricted.Linear
+import Data.Functor.Linear
+import Data.Monoid.Linear
+import Data.Num.Linear
+import Data.List.NonEmpty (NonEmpty ((:|)))
+import GHC.Stack
+import qualified Data.List as NonLinear
+
+-- # Basic functions
+--------------------------------------------------
+
+(++) :: [a] #-> [a] #-> [a]
+(++) = Unsafe.toLinear2 (NonLinear.++)
+
+map :: (a #-> b) -> [a] #-> [b]
+map = fmap
+
+-- | @filter p xs@ returns a list with elements satisfying the predicate.
+--
+-- See 'Data.Maybe.Linear.mapMaybe' if you do not want the 'Dupable' constraint.
+filter :: Dupable a => (a #-> Bool) -> [a] #-> [a]
+filter _ [] = []
+filter p (x:xs) =
+  dup x & \case
+    (x', x'') ->
+      if p x'
+      then x'' : filter p xs
+      else x'' `lseq` filter p xs
+
+uncons :: [a] #-> Maybe (a, [a])
+uncons [] = Nothing
+uncons (x:xs) = Just (x, xs)
+
+reverse :: [a] #-> [a]
+reverse = Unsafe.toLinear NonLinear.reverse
+
+-- | Return the length of the given list alongside with the list itself.
+length :: [a] #-> ([a], Ur Int)
+length = Unsafe.toLinear $ \xs ->
+  (xs, Ur (NonLinear.length xs))
+-- We can only do this because of the fact that 'NonLinear.length'
+-- does not inspect the elements.
+
+--  'splitAt' @n xs@ returns a tuple where first element is @xs@ prefix of
+-- length @n@ and second element is the remainder of the list.
+splitAt :: Int -> [a] #-> ([a], [a])
+splitAt i = Unsafe.toLinear (Prelude.splitAt i)
```

Comment:
Okay, I think you should just make an issue for writing out the implementation for anything from `Data.List` that isn't optimized and not worry about it in this PR.